src/v/cloud_topics/objects/level_one_object.cc

```diff
@@ -0,0 +1,442 @@
+          });
+        if (index_it == rev.end()) {
+            return partition.file_position;
+        }
+        return index_it->file_position;
+    }
+    if (min_partition_after_target != partitions.end()) {
+        return min_partition_after_target->file_position;
+    }
+    return npos;
+}
+
+size_t object_index::file_position_before_max_timestamp(
+  const model::ntp& ntp, model::timestamp target) {
+    auto filtered = std::views::filter(
+      partitions, [&ntp, &target](const auto& p) {
+          return p.ntp == ntp && target <= p.max_timestamp;
+      });
+    auto min_it = std::ranges::min_element(
+      filtered, std::less<>{}, [](const auto& p) { return p.first_offset; });
+    if (min_it == filtered.end()) {
+        return npos;
+    }
+    const auto& index = min_it->indexes;
+    auto it = std::ranges::lower_bound(
+      index, target, std::less<>{}, [](const auto& entry) {
+          return entry.max_timestamp;
+      });
+    // If we're past all index entries, but still within the recorded file
+    // bounds, the best we can do is start at the last well known offset (the
+    // last index entry).
+    if (it == index.end()) {
+        return min_it->indexes.back().file_position;
+    }
+    // If at the first entry, we must start at the file beginning, because the
+    // the max is inclusive of those entries.
+    if (it == index.begin()) {
+        return min_it->file_position;
+    }
+    return it->file_position;
+}
+
+object_index object_index::copy() const {
+    object_index copy;
+    copy.partitions.reserve(partitions.size());
+    for (const auto& p : partitions) {
+        copy.partitions.push_back(p.copy());
+    }
+    return copy;
+}
+
+std::variant<object_index, size_t> object_index::read_footer(iobuf buf) {
+    if (buf.size_bytes() < sizeof(uint32_t)) {
+        throw std::runtime_error(fmt::format(
+          "expected at least {} bytes in footer, got: {}",
+          sizeof(uint32_t),
+          buf.size_bytes()));
+    }
+    iobuf_const_parser parser(buf);
+    auto footer_size
+      = iobuf_parser(buf.tail(sizeof(uint32_t))).consume_type<uint32_t>();
+    if (buf.size_bytes() >= (footer_size + sizeof(uint32_t))) {
+        iobuf_parser p(buf.share(
+          buf.size_bytes() - sizeof(uint32_t) - footer_size, footer_size));
+        auto dt = static_cast<data_type>(
+          p.consume_type<std::underlying_type_t<data_type>>());
+        if (dt != data_type::footer) {
+            throw std::runtime_error(fmt::format(
+              "expected footer data type, got: {}", std::to_underlying(dt)));
+        }
+        auto size = p.consume_type<uint32_t>();
+        if (size != p.bytes_left()) {
+            throw std::runtime_error(fmt::format(
+              "expected footer size to match the remaining bytes, "
+              "got: {}, expected: {}",
+              p.bytes_left(),
+              size));
+        }
+        return serde::read<object_index>(p);
```

Comment:
unclear how big these are in practice and if need futurized versions

Reply:
I'll use futurized versions to be safe, I don't think the extra coroutine here will matter from a perf prospective because this is going to mean you're doing cloud IO.

Reply:
Addressed in https://github.com/redpanda-data/redpanda/pull/26563/commits/e11ef1f7c9dc1f9241f76766626166921769f168